### PR TITLE
JAMES-2526 Define the health check API

### PR DESCRIPTION
**Pull request by aduprat**
To be continued: bfc96c9 -> need a way to play authentication only on given paths.

also the documentation on healthcheck endpoint is missing.

### CODE REVIEW

File: core/src/main/java/org/apache/james/core/healthcheck/HealthCheck.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.core.healthcheck;
+
+public interface HealthCheck {
+
```
Comment (chibenwa) [inline core/src/main/java/org/apache/james/core/healthcheck/HealthCheck.java]: I would find it cool to have a component name as part of the API

File: core/src/main/java/org/apache/james/core/healthcheck/ResultStatus.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.core.healthcheck;
+
+public enum ResultStatus {
+    HEALTHY, UNHEALTHY;
```
Comment (chibenwa) [inline core/src/main/java/org/apache/james/core/healthcheck/ResultStatus.java]: I'd suggest an intermediate `DEGRADED` state: the service is usable but slow...

File: core/src/test/java/org/apache/james/core/healthcheck/ResultTest.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    @Test
+    public void statusShouldBeUnhealthyWhenUnhealthy() {
+        Result result = Result.unhealthy("cause");
+
+        assertThat(result.getStatus()).isEqualTo(ResultStatus.UNHEALTHY);
+    }
+
+    @Test
+    public void causeMayBeEmptyWhenUnhealthy() {
+        Result result = Result.unhealthy(null);
```
Comment (chibenwa) [inline core/src/test/java/org/apache/james/core/healthcheck/ResultTest.java]: IMO this should throw and we should have a `Result.unhealthy()` if not having causes is supported.

File: server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/routes/HealthCheckRoutes.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            @ApiResponse(code = HttpStatus.INTERNAL_SERVER_ERROR_500,
+                message = "Internal server error - When one check has failed.")
+    })
+    public void validateHealthchecks() {
+        service.get(HEALTHCHECK,
+            (request, response) -> {
+                List<Result> anyUnhealthy = retrieveUnhealthyHealthChecks();
+
+                if (!anyUnhealthy.isEmpty()) {
+                    anyUnhealthy.stream()
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/routes/HealthCheckRoutes.java]: .stream is redoundant and not needed for `forEach`

File: server/protocols/webadmin/webadmin-core/src/test/java/org/apache/james/webadmin/routes/HealthCheckRoutesTest.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
+/****************************************************************
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-core/src/test/java/org/apache/james/webadmin/routes/HealthCheckRoutesTest.java]: I'd suggest to extract healthy/unhelthy checks as constants

File: core/src/main/java/org/apache/james/core/healthcheck/HealthCheck.java (hunk 1 above)
Comment (mbaechler) [reply core/src/main/java/org/apache/james/core/healthcheck/HealthCheck.java]: I agree

File: core/src/main/java/org/apache/james/core/healthcheck/ResultStatus.java (hunk 2 above)
Comment (mbaechler) [reply core/src/main/java/org/apache/james/core/healthcheck/ResultStatus.java]: slow or anything else, yes

File: core/src/test/java/org/apache/james/core/healthcheck/ResultTest.java (hunk 3 above)
Comment (mbaechler) [reply core/src/test/java/org/apache/james/core/healthcheck/ResultTest.java]: I don't understand this comment. What do you want to throw ?

Comment (aduprat) [reply core/src/test/java/org/apache/james/core/healthcheck/ResultTest.java]: Having an _unhealthy_ method without parameter & throw in the method with parameter when the String is null.

File: server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/routes/HealthCheckRoutes.java [hunk 6]
```diff
@@ -80,10 +82,10 @@ public void define(Service service) {
     public void validateHealthchecks() {
         service.get(HEALTHCHECK,
             (request, response) -> {
-                List<Pair<ComponentName, Result>> anyUnhealthy = retrieveUnhealthyHealthChecks();
+                List<Pair<ComponentName, Result>> anyUnhealthyOrDegraded = retrieveUnhealthyOrDegradedHealthChecks();
```
Comment (aduprat) [inline server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/routes/HealthCheckRoutes.java]: As there's no content in the response yet, I prefer to answer 500 when a check is degrade.
This way the user will have to check the logs in order to know what's happening.

File: core/src/main/java/org/apache/james/core/healthcheck/ComponentName.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.core.healthcheck;
+
+public class ComponentName {
```
Comment (mbaechler) [inline core/src/main/java/org/apache/james/core/healthcheck/ComponentName.java]: that would make sense to put the ComponentName into Result object

File: server/protocols/webadmin/webadmin-core/src/test/java/org/apache/james/webadmin/authentication/MockAuthenticationFilter.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.webadmin.authentication;
+
+import spark.Request;
+import spark.Response;
+
+public class MockAuthenticationFilter implements AuthenticationFilter {
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-core/src/test/java/org/apache/james/webadmin/authentication/MockAuthenticationFilter.java]: it's not a mock, it's a spy

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/DomainQuotaRoutesTest.java [hunk 9]
```diff
@@ -60,6 +61,73 @@ void setUp(WebAdminQuotaSearchTestSystem testSystem) throws Exception {
...
+        assertThat(testSystem.getAuthenticationFilter().hasBeenAuthenticated()).isTrue();
+    }
+
+    @Test
+    void putCountShouldBeAuthenticated(WebAdminQuotaSearchTestSystem testSystem) {
+        with()
+            .body("42")
+            .put(QUOTA_DOMAINS + "/" + PERDU_COM + "/" + COUNT);
+
+        assertThat(testSystem.getAuthenticationFilter().hasBeenAuthenticated()).isTrue();
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/DomainQuotaRoutesTest.java]: usually, we just try without authentication and check that return code is 401

**Comment by aduprat** [comment]
push -f

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/DomainQuotaRoutesTest.java (hunk 9 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/DomainQuotaRoutesTest.java]: Agree. JWT set up in WebAdmin route tests is a must.

Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/DomainQuotaRoutesTest.java]: Then the downside is that our testing code do not act only on the HTTP layer, and thus, is less portable.

Comment (aduprat) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/DomainQuotaRoutesTest.java]: Yes, but you are not locked on a bounded authentication filter; i.e. changing this filter will break everything.

File: core/src/test/java/org/apache/james/core/healthcheck/ResultTest.java [hunk 10]
```diff
@@ -25,136 +25,160 @@
...
-        Result result = Result.degraded("cause");
+        Result result = Result.degraded(new ComponentName("component"), "cause");
 
         assertThat(result.getStatus()).isEqualTo(ResultStatus.DEGRADED);
     }
 
     @Test
     public void causeMayBeEmptyWhenDegraded() {
-        Result result = Result.degraded();
+        Result result = Result.degraded(new ComponentName("component"));
```
Comment (chibenwa) [inline core/src/test/java/org/apache/james/core/healthcheck/ResultTest.java]: We can use a constant for `new ComponentName("component")` ?

File: server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/routes/HealthCheckRoutes.java [hunk 11]
```diff
@@ -82,44 +80,44 @@ public void define(Service service) {
...
 
-    private void logFailedCheck(Pair<ComponentName, Result> pair) {
-        switch (pair.getValue().getStatus()) {
+    private void logFailedCheck(Result result) {
+        switch (result.getStatus()) {
         case UNHEALTHY:
             LOGGER.error("HealthCheck failed for {} : {}",
-                    pair.getKey().getName(),
-                    pair.getValue().getCause().orElse(""));
+                    result.getComponentName().toString(),
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/routes/HealthCheckRoutes.java]: Why not keep getName ?

File: server/protocols/webadmin/webadmin-core/src/test/java/org/apache/james/webadmin/routes/HealthCheckRoutesTest.java [hunk 12]
```diff
@@ -135,8 +135,8 @@ public void validateHealthchecksShouldReturnInternalErrorWhenAllHealthChecksAreU
 
     @Test
     public void validateHealthchecksShouldReturnInternalErrorWhenOneHealthCheckIsDegraded() {
-        healthChecks.add(healthCheck(new ComponentName("component-1"), Result.degraded("cause")));
-        healthChecks.add(healthCheck(new ComponentName("component-2"), Result.healthy()));
+        healthChecks.add(healthCheck(Result.degraded(new ComponentName("component-1"), "cause")));
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-core/src/test/java/org/apache/james/webadmin/routes/HealthCheckRoutesTest.java]: Constant extraction for component name would make sens here too

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
I split this PR in 2 blocks:

 - https://github.com/linagora/james-project/pull/1690 for the health check part
 - https://github.com/linagora/james-project/pull/1691 for the public webaedmin endpoints

**Pull request closed without merging**